kernel: kconfig: Disable DEVICE_DEINIT_SUPPORT by default

nordicjm opened the pull request:
This Kconfig, which it itself admits is for a "very specific case" was set to default as yes, this includes extra code in drivers with this functionality and increases driver struct size for cases where this function isn't needed (i.e. all because it's enabled by default), therefore change it to be opt-in rather than opt-out

Depends on https://github.com/zephyrproject-rtos/zephyr/pull/99391

nordicjm committed:
kernel: kconfig: Disable DEVICE_DEINIT_SUPPORT by default

This Kconfig, which it itself admits is for a "very specific case"
was set to default as yes, this includes extra code in drivers with
this functionality and increases driver struct size for cases where
this function isn't needed (i.e. all because it's enabled by
default), therefore change it to be opt-in rather than opt-out

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

peter-mitsis commented:
I'm ok with setting this as disabled by default. It just looks like there are some drivers that will need to be tidied up as there are now instances of functions being defined but no longer used.

nordicjm committed:
drivers: serial: uart_bflb: Fix not guarding deinit function

Fixes an issue with an unused function being defined if
CONFIG_DEVICE_DEINIT_SUPPORT was disabled

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: kernel: device: Add CONFIG_DEVICE_DEINIT_SUPPORT

Adds this Kconfig, which is needed for the test, because it will
no longer be globally enabled by default

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
tests: pm: device_driver_init: Add CONFIG_DEVICE_DEINIT_SUPPORT

Adds this Kconfig, which is needed for the test, because it will
no longer be globally enabled by default

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
doc: services: pm: device: Fix missing #ifdef

Adds a missing #ifdef for CONFIG_DEVICE_DEINIT_SUPPORT for the
driver deinit function example

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>

nordicjm committed:
drivers: i2c: i2c_bflb: Fix not guarding deinit function

Fixes an issue with an unused function being defined if
CONFIG_DEVICE_DEINIT_SUPPORT was disabled

Signed-off-by: Jamie McCrae <jamie.mccrae@nordicsemi.no>
